core/src/main/java/jenkins/tasks/SimpleBuildStep.java

```diff
@@ -80,9 +82,37 @@
      * @param listener a place to send output
      * @throws InterruptedException if the step is interrupted
      * @throws IOException if something goes wrong; use {@link AbortException} for a polite error
+     * @deprecated Use {@link #perform(Run, FilePath, EnvVars, Launcher, TaskListener) instead.}
```

Comment:
There are a couple of implementations inside core which should be adapted, though they would ignore the `EnvVars` argument.

Reply:
I was more thinking of adapting them to better follow the substitution guidelines for steps, i.e. not even apply substitution based on the initial environment. But I left that out because it didn't seem directly related to the ticket at hand.
But I'll add a commit that adapts them to the new API.

Reply:
> I was more thinking of adapting them to better follow the substitution guidelines for steps, i.e. not even apply substitution based on the initial environment.

Oh maybe you are thinking about https://github.com/jenkinsci/jenkins/blob/449c5aced523a6e66fe3d6a804e5dbfd5c5c67c6/core/src/main/java/hudson/tasks/ArtifactArchiver.java#L248 & https://github.com/jenkinsci/jenkins/blob/d30edab1746555a22a99b17460a36ea2df949e68/core/src/main/java/hudson/tasks/Fingerprinter.java#L177-L179 and are correct that they _would_ use the `EnvVars` argument—but you can do so only if `build instanceof AbstractBuild`, which will skip expansion for `WorkflowRun` like we want. See [JENKINS-35671](https://issues.jenkins-ci.org/browse/JENKINS-35671).